docs/proposals/petset.md

```diff
@@ -0,0 +1,414 @@
+  focus on ensuring the information necessary for the software to function is available. For example,
+  rather than providing a "post-creation" hook invoked when the cluster is complete, provide the necessary
+  information to the "first" (or last) pod to determine the identity of the remaining cluster members and
+  allow it to manage its own initialization.
+* **External access direct to cluster members is out of scope** - exposing pods to consumers that cannot
+  access the pod network is out of scope, because we do not currently support headless services being
+  exposed via NodePort. A workaround is to allow external clients to access the pod network, or to
+  create one NodePort service per member. A future design should cover headless external service access.
+
+
+## Proposed Design
+
+Add a new resource to Kubernetes to represent a set of pods that are individually distinct but each
+individual can safely be replaced-- the name **PetSet** (working name) is chosen to convey that the
+individual members of the set are themselves "pets" and thus each one is preserved. A relevant analogy
+is that a PetSet is composed of pets, but the pets are like goldfish. If you have a blue, red, and
+yellow goldfish, and the red goldfish dies, you replace it with another red goldfish and no one would
+notice. If you suddenly have three red goldfish, someone will notice.
+
+The PetSet is responsible for creating and maintaining a set of **identities** and ensuring that there is
+one pod and zero or more **supporting resources** for each identity. There should never be more than one pod
+or unique supporting resource per identity at any one time. A new pod can be created for an identity only
+if a previous pod has been fully terminated (reached its graceful termination limit or cleanly exited).
+
+A PetSet has 0..N **members**, each with a unique **identity** which is a name that is unique within the
+set.
+
+```
+type PetSet struct {
+    ObjectMeta
+
+    Spec PetSetSpec
+    ...
+}
+
+type PetSetSpec struct {
+		// A label selector that "owns" objects created under this set
+		Selector map[string]string
+
+    // If set, identities will be auto generated based on the name of the
+    // Set "NAME-$(identity)", i.e. "mysql-1", "mysql-2", ...
+    Members *int
+    // A structured set of identities that if set will be used to generate more
+    /// complex shard patterns
+    MemberIdentities []MemberIdentity
+}
+
+type MemberIdentity struct {
+    // a set of fixed identities, like ["mysql-master", "mysql-slave-1"]
+    Fixed []string
+    // a pattern for a range of identities
+    Range *IdentityRange
+    // a set of nested identities that are constructed from the parent pattern,
+    // allows nesting like "mysql-zone1-1", "mysql-zone1-2", "mysql-zone2-1"
+    Identities []MemberIdentity
+
+    // PROPOSED: Add a set of additional parameters per identity that can be
+		// injected into the templates
+}
```

Comment:
> Since you can create disjoint sets by specifying different named parameters in the template, do we really need (min, max)? 

Yes, for autoscaling.
But I'm still stuck at: do we need the nested identities vs something like: "foo-$(X1)-zone-$(X2)-slave-shard-$(X3)" in the template coupled with:
```
parameters:
  - name: X1
    value: [1..N]
  - name: X2
    value: [4..N]
```

That does make autoscaling hard because I don't know which of X 1,2,3 does what.